astropy/time/formats.py

```diff
@@ -720,11 +712,33 @@ def parse_string(self, timestr, subfmts):
                 if tm is None:
                     continue
                 tm = tm.groupdict()
+                if '.' in tm.get('sec', ''):
```

Comment:
I wonder if you can tidy this up a bit by making the fractional part a named component.  E.g.
```
In [24]: r = r'(?P<sec>\d{1,2})(?P<fracsec>\.\d*)?'

In [25]: m = re.match(r, '12.1')

In [26]: m.groupdict()
Out[26]: {'fracsec': '.1', 'sec': '12'}

In [27]: m = re.match(r, '12.')

In [28]: m.groupdict()
Out[28]: {'fracsec': '.', 'sec': '12'}

In [29]: m = re.match(r, '12')

In [30]: m.groupdict()
Out[30]: {'fracsec': None, 'sec': '12'}
```
Then:
```
fracsec = tm.get('fracsec')
if fracsec is not None:
    fracsec = float('0' + fracsec)
```